### ISSUE TITLE
[nix-local-build] unbounded-delays Setup on GHC 8.0rc2 causing problems for nix-local-build

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#3248]
When I attempt to new-build the nix-local-branch with GHC 8.0rc2 I get the following error:

```
ezyang@sabre:~/Dev/cabal-nix-local-build$ cabal new-build  -w ghc-8.0 --enable-tests
Resolving dependencies...
cabal: Could not resolve dependencies:
trying: Cabal-test.tasty-quickcheck~>cabal-install-test.tasty-quickcheck-0.8.4
(dependency of cabal-install-test.Cabal-1.24.0.0:*test)
trying: Cabal-test.tasty-quickcheck-0.8.4:!test
trying: Cabal-test.tasty-hunit~>cabal-install-test.tasty-hunit-0.9.2
(dependency of cabal-install-test.Cabal-1.24.0.0:*test)
trying: Cabal-test.tasty~>cabal-install-test.tasty-0.11.0.2 (dependency of
cabal-install-test.Cabal-1.24.0.0:*test)
trying:
Cabal-test.unbounded-delays~>cabal-install-test.unbounded-delays-0.1.0.9
(dependency of Cabal-test.tasty-0.11.0.2)
trying: Cabal-test.tasty-quickcheck-0.8.4:-old-quickcheck
trying: unbounded-delays-setup.Cabal~>cabal-install-test.Cabal-1.24.0.0
(dependency of cabal-install-test.unbounded-delays-0.1.0.9)
rejecting: unbounded-delays-setup.Cabal-1.24.0.0:!test (constraint from config
file, command line flag, or user target requires opposite flag selection)
trying: unbounded-delays-setup.Cabal-1.24.0.0:*test
trying: unbounded-delays-setup.Cabal-1.24.0.0:-bundled-binary-generic
Backjump limit reached (currently 2000, change with --max-backjumps or try to
run with --reorder-goals).
```

This doesn't happen if I use GHC 7.10. I don't really know how to read this error message, which is another problem in and of itself.

I can "solve" the problem by vendoring unbounded-delays to have a non-Custom setup script.

**Comment by ezyang** [comment]
I think I know what the problem is: unbounded-delays has a qualified setup goal which depends on Cabal; unfortunately, when we say `--enable-tests`, the solver tries to apply this flag to the qualified goal, even when it shouldn't. I'll try to come up with a fix.

**Comment by ezyang** [comment]
It turns out, the solver was correctly not applying the constraint to qualified goals; however, the setup dependency was getting *linked* to the primary dependency, leading the constraint to be transitively applied. Disaster! #3268 fixes this.

**Comment by grayjay** [comment]
Was this fixed by #3357?

**Comment by ezyang** [comment]
Yes, thanks!
